File: web_version/pmugg_files/network/vertexPlacement.js
ms.vertexPlacement = function(vertex, id, settings) {
	this.vertex = vertex;
	this.id = id;
	this.settings = settings;
	this.freeFaceIds = [];
	this.unfreeFaceIds = [];
	// This ID depends on the others. It cannot be used as one of the three unfreeFaceIds.
	this.colinearFaceIds = [];

	this.M = null;

	// How the position varies with the current range we are examining.
	// When the vertex position is set, these should be null and position should be used
	this.slope = null;
	this.value = null;
};

ms.vertexPlacement.prototype.initialize = function() {
	var endpoints = this.vertex.getEndpoints();
	var self = this;
	var { settings } = this;
	endpoints.forEach((endpoint) => {
		var face = endpoint.getFace();
		var id = face.getNode().id;
		if (!settings.getFace(id)) {
			var normal = face.getFaceType().getNormal();
			settings.facePlacements[id] = new ms.facePlacement(normal, id, settings, face);
		}
		this.addFreeFace(id);
	});
};

// Add a fixed neighbor to all the faces.
ms.vertexPlacement.prototype.addFixedNeighbor = function(fixedFace) {
	var { settings } = this;
	this.getAllFaceIds().forEach((id) => {
		if (id != fixedFace.fPlace.id) {
			settings.getFace(id).addFixedNeighbor(fixedFace.faceA);
		}
	});
};

ms.vertexPlacement.prototype.addFreeFace = function(id) {
	var settings = this.settings;
	var newId = settings.uniqueFaceMap[id];
	if (newId !== undefined) { id = newId; }
	if (this.freeFaceIds.includes(id)) {
		return;
	}
	var newFreeFace = this.settings.getFace(id);
	var coplanarId = this.freeFaceIds.find((id) => {
		return newFreeFace.coplanar(settings.getFace(id));
	});
	if (coplanarId) {
		settings.mergeFace(coplanarId, id);
		id = coplanarId;
	} else {
		this.freeFaceIds.push(id);	
	}
	
	this.settings.getFace(id).addVertexId(this.id);
	
	var colinear = (this.freeFaceIds.length == 3 && !this.getA(this.freeFaceIds));
	if (colinear) {
		this.freeFaceIds.pop();
	}
};

// Check if there are three faces. Add one if there is not.
ms.vertexPlacement.prototype.checkThreeFaces = function(id) {
	var { settings } = this;
	if (this.freeFaceIds.length < 2) {
		ms.alert('Vertex should have at least two faces.');
	} else if (this.freeFaceIds.length == 2) {
		var n0 = settings.getFace(this.freeFaceIds[0]).getNormal();
		var n1 = settings.getFace(this.freeFaceIds[1]).getNormal();
		var n2 = n0.cross(n1);
		n2.normalize();
		var id = this.settings.createFace(n2);
		this.addFreeFace(id);
	}
};

ms.vertexPlacement.prototype.getNumConstraints = function() {
	return this.unfreeFaceIds.length;
};

ms.vertexPlacement.prototype.getPosition = function() {
	return this.value;
};

ms.vertexPlacement.prototype.getAllFaceIds = function() {
	return this.unfreeFaceIds.concat(this.freeFaceIds, this.colinearFaceIds);
};

ms.vertexPlacement.prototype.fixPosition = function() {
	var { settings } = this;
	this.slope = ms.vec3.ORIGIN;
	this.value = this.vertex.getPosition();
	var self = this;
	var success = true;
	this.getAllFaceIds().forEach((id) => {
		success = success && settings.getFace(id).setFromVertex(self.id);
		settings.getFace(id).setFixed(true);
	});
	return success;
};

ms.vertexPlacement.prototype.addConstraint = function() {
	var { settings } = this;
	var freeFaceId = this.freeFaceIds[0];
	settings.getFace(freeFaceId).constrain(true);
};

ms.vertexPlacement.prototype.constrainFace = function(id) {
	ms.remove(id, this.freeFaceIds);
	this.unfreeFaceIds.push(id);

	// Check if the face IDs are colinear.
	if (this.unfreeFaceIds.length == 3) {
		var M = this.getM();
		if (!M) {
			this.unfreeFaceIds.pop();
			this.colinearFaceIds.push(id);
		}
	}
};

ms.vertexPlacement.prototype.propagate = function(id) {
	var { settings } = this;
	var self = this;
	if (this.unfreeFaceIds.length >= 3) {
		settings.addToOrder(this.id, 'vertex');
		var freeIds = this.freeFaceIds.concat(this.colinearFaceIds).slice();
		freeIds.forEach((id) => {
			var fPlace = settings.getFace(id);
			if (fPlace.isFree()) {
				fPlace.constrain(false, this.id);
			}
		});
		this.vertex.getEndpoints().forEach((endpoint) => {
			var id = endpoint.getLine().getNode().getId()
			var ePlace = settings.getEdge(id);
			ePlace && ePlace.addConstraint(self.id);
			var prev = endpoint.prev();
			if (prev.getEdgeType().faceData.length == 1) {
				// If prev edge only has one endpoint. Propagate to previous vertex.				
				var id = prev.getLine().getNode().getId()
				var ePlace = settings.getEdge(id);
				ePlace && ePlace.addConstraint(self.id);
			}
		});
	}
};

// Ax = D where A is the normals. a_i *x + b_i * y + c_i * z = d_i.
// A^-1 * D = x. M = A^-1.
// Returns null, if A is not invertible.
ms.vertexPlacement.prototype.getA = function(faceIds) {
	var { settings } = this;
	var A = [];
	for (var i = 0; i < 3; i++) {
		var id = faceIds[i];
		var fPlace = settings.getFace(id);
		var n = fPlace.getNormal();
		A.push([n.x, n.y, n.z]);
	}
	return mathG.matrix(A);
};

// Ax = D where A is the normals. a_i *x + b_i * y + c_i * z = d_i.
// A^-1 * D = x. M = A^-1.
// Returns null, if A is not invertible.
ms.vertexPlacement.prototype.getM = function() {
	if (!this.M) {
		var A = this.getA(this.unfreeFaceIds);
		// The three faces are colinear.
		if (Math.abs(mathG.det(A)) < 1e-8) {
			return null;
		}
		this.M = mathG.inv(A);
	}
	return this.M;
};

// Vertex positions x = M * D.
ms.vertexPlacement.prototype.setPosition = function() {
	var { settings } = this;
	var D = [];
	for (var i = 0; i < 3; i++) {
		var id = this.unfreeFaceIds[i];
		var fPlace = settings.getFace(id);
		D.push([fPlace.getD()]);
	}
	D = mathG.matrix(D);
	var b = mathG.multiply(this.getM(), D).valueOf();
	this.slope = ms.vec3.ORIGIN;
	this.value = new ms.vec3(b[0][0], b[1][0], b[2][0]);
};

// Gets the range of possible values for the vertex.
ms.vertexPlacement.prototype.getRange = function() {
	var self = this;
	var { settings } = this;

	// Figure out how the three unfree faces vary.
	var m3 = [];
	var b3 = [];
	for (var i = 0; i < 3; i++) {
		var id = this.unfreeFaceIds[i];
		var fPlace = settings.getFace(id);
		var mbi = fPlace.getChangeMB();
		m3.push([mbi.m]);
		b3.push([mbi.b]);
	}
	m3 = mathG.matrix(m3);
	b3 = mathG.matrix(b3);
	var m = mathG.multiply(this.getM(), m3).valueOf();
	var b = mathG.multiply(this.getM(), b3).valueOf();

	var {lower, upper} = settings;
	var range = new ms.range(-Infinity, Infinity);
	for (var i = 0; i < 3; i++) {
		var rangeI = ms.range.transformCreate(m[i][0], b[i][0], new ms.range(lower[i], upper[i]));
		range = range.intersect(rangeI);
	}

	this.slope = new ms.vec3(m[0][0], m[1][0], m[2][0]);
	this.value = new ms.vec3(b[0][0], b[1][0], b[2][0]);

	return range;
};

ms.vertexPlacement.prototype.getChangeMB = function() {
	return {m: this.slope, b: this.value};
};

ms.vertexPlacement.prototype.print = function() {
	this.vertex.print();
};

File: cpp_version/pmugg/network/vertex_placement.cpp
#include "pch.h"
#include "vertex_placement.h"

namespace ms {

void VertexPlacement::initialize() {
    std::vector<Endpoint*> endpoints = vertex->getEndpoints();

    for (auto* endpoint : endpoints) {
        if (!endpoint) {
            // Happens in the ground transition.
            continue;
        }
        Face* face = endpoint->getFace();
        int id = face->getId();

        if (!settings->getFace(id)) {
            auto normal = face->getFaceType()->getNormal();
            settings->facePlacements[id] = std::make_unique<FacePlacement>(normal, id, settings, face);
        }
        addFreeFace(id);
    }
}

void VertexPlacement::addConstraint() {
    if (freeFaceIds.empty()) {
        throw std::runtime_error("No free faces available for constraint");
    }
    int freeFaceId = freeFaceIds[0];
    settings->getFace(freeFaceId)->constrain(true, -1);
}

void VertexPlacement::constrainFace(int id) {
    auto it = std::find(freeFaceIds.begin(), freeFaceIds.end(), id);
    if (it != freeFaceIds.end()) {
        freeFaceIds.erase(it);
    }
    unfreeFaceIds.push_back(id);

    // Check if the face IDs are colinear
    if (unfreeFaceIds.size() == 3) {
        auto* M = getM();
        if (!M) {
            unfreeFaceIds.pop_back();
            colinearFaceIds.push_back(id);
        }
    }
}

void VertexPlacement::propagate() {
    if (unfreeFaceIds.size() >= 3) {
        settings->addToOrder(this->id, "vertex", -1);
        
        // Handle free faces and colinear faces
        auto freeIds = freeFaceIds;
        freeIds.insert(freeIds.end(), colinearFaceIds.begin(), colinearFaceIds.end());
        
        for (int faceId : freeIds) {
            auto* fPlace = settings->getFace(faceId);
            if (fPlace->isFree()) {
                fPlace->constrain(false, this->id);
            }
        }

        // Handle connected edges
        for (Endpoint* endpoint : vertex->getEndpoints()) {
            int edgeId = endpoint->getLine()->getId();
            auto* ePlace = settings->getEdge(edgeId);
            if (ePlace) {
                ePlace->addConstraint(this->id);
            }

            auto* prev = endpoint->prev();
            if (prev->getEdgeType()->faceData.size() == 1) {
                // If prev edge only has one endpoint, propagate to previous vertex
                int prevEdgeId = prev->getLine()->getId();
                auto* prevEPlace = settings->getEdge(prevEdgeId);
                if (prevEPlace) {
                    prevEPlace->addConstraint(this->id);
                }
            }
        }
    }
}

Matrix* VertexPlacement::getA(const std::vector<int>& faceIds) {
    std::vector<std::vector<double>> A(3, std::vector<double>(3)); // Assuming A is a 3x3 matrix

    for (size_t i = 0; i < 3; i++) {
        int id = faceIds[i];
        FacePlacement* fPlace = settings->getFace(id);
        Vec3 n = fPlace->getNormal();
        A[i] = { n.getX(), n.getY(), n.getZ()}; // Fill the matrix row with the normal components
    }

    return new Matrix(A);
}

Matrix* VertexPlacement::getM() {
    if (!M) {
        Matrix* A = getA(unfreeFaceIds);
        // The three faces are collinear.
        if (std::abs(Matrix::det(*A)) < 1e-8) {
            delete A; // Clean up memory
            return nullptr; // Return null if A is not invertible
        }
        M = Matrix::inverse(A);
        delete A; // Clean up memory for A
    }
    return M;
}

void VertexPlacement::setPosition() {
    std::vector<std::vector<double>> D(3, std::vector<double>(1)); // 3x1 matrix

    for (size_t i = 0; i < 3; i++) {
        int id = unfreeFaceIds[i];
        FacePlacement* fPlace = settings->getFace(id);
        D[i][0] = fPlace->getD(); // Fill D with D values
    }

    Matrix* DMatrix = new Matrix(D); // Create a matrix from D
    auto bMatrix = Matrix::multiply(getM(), DMatrix); // Multiply M and D
    auto b = bMatrix->valueOf(); // Multiply M and D
    delete DMatrix; // Clean up memory for DMatrix

    slope = Vec3(0, 0, 0); // Initialize slope
    value = Vec3(b[0][0], b[1][0], b[2][0]); // Assuming b is a 3x1 matrix
    delete bMatrix; // Clean up memory for bMatrix
}

Range VertexPlacement::getRange() {
    std::vector<std::vector<double>> m3(3, std::vector<double>(1));
    std::vector<std::vector<double>> b3(3, std::vector<double>(1));

    for (size_t i = 0; i < 3; i++) {
        int id = unfreeFaceIds[i];
        FacePlacement* fPlace = settings->getFace(id);
        auto mbi = fPlace->getChangeMB();
        m3[i][0] = mbi.m; // Assuming mbi.m is a Vec3 or similar
        b3[i][0] = mbi.b; // Assuming mbi.b is a double
    }

    Matrix* m3Matrix = new Matrix(m3);
    Matrix* b3Matrix = new Matrix(b3);
    Matrix* mMatrix = Matrix::multiply(getM(), m3Matrix);
    Matrix* bMatrix = Matrix::multiply(getM(), b3Matrix);
    auto m = mMatrix->valueOf();
    auto b = bMatrix->valueOf();

    delete m3Matrix; // Clean up memory for m3Matrix
    delete b3Matrix; // Clean up memory for b3Matrix

    Range range(-std::numeric_limits<double>::infinity(), std::numeric_limits<double>::infinity());
    for (size_t i = 0; i < 3; i++) {
        Range rangeI = Range::transformCreate(m[i][0], b[i][0], Range(settings->lower[i], settings->upper[i]));
        range = range.intersect(rangeI);
    }

    slope = Vec3(m[0][0], m[1][0], m[2][0]);
    value = Vec3(b[0][0], b[1][0], b[2][0]);

    delete mMatrix; // Clean up memory for m
    delete bMatrix; // Clean up memory for b

    return range;
}

ChangeVecMB VertexPlacement::getChangeMB() const {
    auto m = slope;
    auto b = value;
    return { m, b };
}

const Vec3& VertexPlacement::getPosition() const {
    return value; // Return the position value
}

void VertexPlacement::addFixedNeighbor(const FixedFace& fixedFace) {
    auto faceIds = getAllFaceIds(); // Get all face IDs
    for (int id : faceIds) {
        if (id != fixedFace.fPlace->id) { // Assuming Face has a method getId()
            settings->getFace(id)->addFixedNeighbor(fixedFace.faceA); // Assuming faceA is accessible
        }
    }
}

void VertexPlacement::addFreeFace(int id) {
    auto it = settings->uniqueFaceMap.find(id);
    if (it != settings->uniqueFaceMap.end()) {
        id = it->second; // Update id if a new ID is found
    }

    // Check if the ID is already in freeFaceIds
    if (std::find(freeFaceIds.begin(), freeFaceIds.end(), id) != freeFaceIds.end()) {
        return; // ID already exists, exit the function
    }

    FacePlacement* newFreeFace = settings->getFace(id);
    int coplanarId = 0;
    bool isCoplanar = false;
    for (int existingId : freeFaceIds) {
        FacePlacement* existingFace = settings->getFace(existingId);
        if (newFreeFace->coplanar(existingFace)) {
            coplanarId = existingId;
            isCoplanar = true;
            break;
        }
    }

    if (isCoplanar) {
        settings->mergeFace(coplanarId, id);
        id = coplanarId; // Update id to the merged face ID
    } else {
        freeFaceIds.push_back(id); // Add the new face ID
    }

    settings->getFace(id)->addVertexId(this->id); // Add the vertex ID to the face

    // Check for colinearity
    //if (freeFaceIds.size() == 3 && getA(freeFaceIds).empty()) { // Assuming getA returns a vector
    //    freeFaceIds.pop_back(); // Remove the last face ID
    //}
}

std::vector<int> VertexPlacement::getAllFaceIds() const {
    std::vector<int> allFaceIds; // Vector to hold all face IDs

    // Reserve space to avoid multiple allocations
    allFaceIds.reserve(unfreeFaceIds.size() + freeFaceIds.size() + colinearFaceIds.size());

    // Concatenate the vectors
    allFaceIds.insert(allFaceIds.end(), unfreeFaceIds.begin(), unfreeFaceIds.end());
    allFaceIds.insert(allFaceIds.end(), freeFaceIds.begin(), freeFaceIds.end());
    allFaceIds.insert(allFaceIds.end(), colinearFaceIds.begin(), colinearFaceIds.end());

    return allFaceIds; // Return the combined vector
}

// Method to get the number of constraints
int VertexPlacement::getNumConstraints() const {
    return unfreeFaceIds.size();
}


void VertexPlacement::checkThreeFaces() {
    if (freeFaceIds.size() < 2) {
        std::cerr << "Vertex should have at least two faces.\n";
    } else if (freeFaceIds.size() == 2) {
        Vec3 n0 = settings->getFace(freeFaceIds[0])->getNormal();
        Vec3 n1 = settings->getFace(freeFaceIds[1])->getNormal();
        Vec3 n2 = n0.cross(n1);
        n2.normalize();
        int newId = settings->createFace(n2);
        addFreeFace(newId);
    }
}

bool VertexPlacement::fixPosition() {
    this->slope = Vec3::ORIGIN;
    this->value = this->vertex->getPosition();

    bool success = true;
    for (const auto& id : this->getAllFaceIds()) {
        success = success && settings->getFace(id)->setFromVertex(this->id);
        settings->getFace(id)->setFixed(true);
    }

    return success;
}

} // namespace ms 